src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -3636,3 +3636,98 @@ mcd_rpc_message_decompress_if_necessary (mcd_rpc_message *rpc, void **data, size
 
    return mcd_rpc_message_decompress (rpc, data, data_len);
 }
+
+bool
+mongoc_cluster_run_retryable_write (mongoc_cluster_t *cluster,
+                                    mongoc_cmd_t *cmd,
+                                    bool is_retryable_write,
+                                    mongoc_server_stream_t **retry_server_stream,
+                                    bson_t *reply,
+                                    bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (cluster);
+   BSON_ASSERT_PARAM (cmd);
+   BSON_ASSERT_PARAM (retry_server_stream);
+   BSON_ASSERT_PARAM (reply);
+
+   bool ret;
+   // `can_retry` is set to false on retry. A retry may only happen once.
+   bool can_retry = is_retryable_write;
+
+   /* increment the transaction number for the first attempt of each retryable
+    * write command */
+   if (is_retryable_write) {
+      bson_iter_t txn_number_iter;
+      BSON_ASSERT (bson_iter_init_find (&txn_number_iter, cmd->command, "txnNumber"));
+      bson_iter_overwrite_int64 (&txn_number_iter, ++cmd->session->server_session->txn_number);
+   }
+
+   // Store the original error and reply if needed.
+   struct {
+      bson_t reply;
+      bson_error_t error;
+      bool set;
+   } original_error = {.reply = {0}, .error = {0}, .set = false};
```

Comment:
```suggestion
   } original_error = {.reply = {0}, .error = {0}, .set = false};

   // Ensure `*retry_server_stream` is always valid or null.
   *retry_server_stream = NULL;
```
To avoid risks such as the following:
```c
// Uninitialized.
mongoc_server_stream_t *retry_server_stream;

// `retry_server_stream` is *maybe* assigned-to.
const bool ret = mongoc_cluster_run_retryable_write (..., &retry_server_stream, ...);

// *Maybe* undefined behavior (still uninitialized).
if (retry_server_stream) {
  // ...
}
```